Database refactoring, creation and using API for the same

suresh-thelkar opened the pull request:
<!--- Please review your changes in preview mode -->
<!--- Provide a general summary of your changes in the Title above -->

#### Bug/feature Description
* Currently the time taken for various PBS Pro database transactions that take place during the job/reservation life cycle is huge. This is becoming a bottleneck for the overall performance of PBS system. This PR addresses these problems through database refactoring, fine tuning the queries, avoiding unnecessary indexes, sequences etc.

#### Affected Platform(s)
* All

#### Cause / Analysis / Design

* As the degree of normalisation of some of the tables in PBS today is very high we are using denormalisation as a strategy to increase the performance. The idea behind is to add redundant data where we think it helps us the most.
* The technique that we are adopting is using hstore. Hstore is essentially a key-value store directly in postgress. The advantage of hstore is we no need to define any of the keys ahead of time. The obvisous benefit here is the flexibility. Postgress automatically indexes every key and value within the hstore which brings us huge performance benefits.


#### Solution Description
Following is the solution followed.
- Using the above techniques described in the analysis/design section we are denormalising the tables server, queue, resv, job, node, scheduler etc by removing their corresponding attributes tables and embedding the information present in those attribute tables in the main tables itself with the help of hstore.
- Also made few more changes where in we removed the unncessary indexes like ji_jobid_idx and database sequences and uncessary primary keys.
- Libdb related changes to support hstore
- Server side code changes to adopt new Libdb changes
- Code changes for various PBS configuration start up scripts/files to accomodate these changes.
- PBS commands which directly depends on database like printjob are also refactored to adopt these changes.
- These code changes aim at the perfomance improvement of PBS system. No external functionality has been changed.

Given below are the ER diagrams before and after this change.
[ERDiagramBeforePR.pdf](https://github.com/PBSPro/pbspro/files/2628515/ERDiagramBeforePR.pdf)
[PBSPro_DB_ERDdiagram_AfterPR.pdf](https://github.com/PBSPro/pbspro/files/2628516/PBSPro_DB_ERDdiagram_AfterPR.pdf)





#### Testing logs/output
* Significant amount of testing is done to ensure that all PBS functionality works fine. Following are the logs attached.
[test_job_sorted_per_scheduler.log](https://github.com/PBSPro/pbspro/files/2628531/test_job_sorted_per_scheduler.log)
[TestEquivClass.log](https://github.com/PBSPro/pbspro/files/2628532/TestEquivClass.log)
[TestFairshare.log](https://github.com/PBSPro/pbspro/files/2628533/TestFairshare.log)
[TestMultipleSchedulers.log](https://github.com/PBSPro/pbspro/files/2628534/TestMultipleSchedulers.log)
[TestReservations.log](https://github.com/PBSPro/pbspro/files/2628535/TestReservations.log)
[TestSchedFifo.log](https://github.com/PBSPro/pbspro/files/2628536/TestSchedFifo.log)
[TestSchedulerInterface.log](https://github.com/PBSPro/pbspro/files/2628537/TestSchedulerInterface.log)
[TestSoftWalltime.log](https://github.com/PBSPro/pbspro/files/2628538/TestSoftWalltime.log)
[TestTrillionJobid.log](https://github.com/PBSPro/pbspro/files/2628539/TestTrillionJobid.log)
[TestVerifyLogOutput.log](https://github.com/PBSPro/pbspro/files/2628540/TestVerifyLogOutput.log)

No external interfaces or functionality of PBS has been changed as part of this PR and hence no need of new PTL test cases. Existing PTL test cases and regression are proof of testing for this PR.

Also following tests are done in order to show the performance improvements that happened due to the changes in this PR. Please note that the performance numbers given below are from a development machine. Numbers will vary from system to system. These simple tests are done just to ensure that the database refactoring changes that are done as part of this PR indeed give us the performance benefits.
1. PBS Server restart when 100,000 jobs in queue. Performance improvement is 81%.
[Scenario_1.pdf](https://github.com/PBSPro/pbspro/files/2628670/Scenario_1.pdf)

2. Single Scheduling Cycle: Submit 100 jobs each user for 10 users when scheduling is off and then turn on scheduling. 
[Scenario_2.pdf](https://github.com/PBSPro/pbspro/files/2628588/Scenario_2.pdf)
Percentage of increase for
->Job submission rate is 23%
->Job Run rate is 456%
->Job Throughput 60%
3. Multiple Scheduling Cycles: Submit 100 jobs per each user by having total of 10 users for total of 10 iterations in multiple scheduling cycles when scheduling is on.
[Scenario_3.pdf](https://github.com/PBSPro/pbspro/files/2628667/Scenario_3.pdf)
->Job submission rate is 38%
->Job Run rate is 192%
->Job Throughput 200%

#### Checklist:
<!--- Use the preview button to see the checkboxes/links properly. -->
<!--- Go over the following points, and put an `x` (without spaces around it) in the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] I have joined the **[pbspro community forum](http://community.pbspro.org/)**.
- [x] My pull request contains a **single, signed** commit. See **[setting up gpg signature](https://pbspro.atlassian.net/wiki/display/DG/Signing+Your+Git+Commits)**.
- [x] My code follows the **[coding style](https://pbspro.atlassian.net/wiki/display/DG/Coding+Standards)** of this project.
- [ ] My change requires project documentation. See **[required documentation checklist](https://pbspro.atlassian.net/wiki/display/DG/Checklist+for+Developing+Features+and+Bug+Fixes)** for details.
   - [ ] I have added documentation in the **[project documentation area](https://pbspro.atlassian.net/wiki/display/PD)**.
- [ ] I have added new **PTL test(s) to my commit**. (See **[using PTL for testing](https://pbspro.atlassian.net/wiki/display/DG/Using+PTL+for+Testing)**) *(or)*
   - [ ] I have added  **manual test(s) to this pull request and explained why PTL is not appropriate** for this case.
- [x] All new and existing automated tests have passed. (See **[running automated PTL tests](https://pbspro.atlassian.net/wiki/display/DG/PTL+Quick+Start+Guide)**).
- [x] I have attached **test logs to this pull request** as evidence of testing/verification.


__***For further information please visit the [Developer Guide Home](https://pbspro.atlassian.net/wiki/display/DG/Developer+Guide+Home).***__

suresh-thelkar commented:
@subhasisb/ @Shrini-h / @nithinj 
Inviting all of you to review this PR as you are aware of the database refactoring changes that are done here.

mike0042 commented on pbspro.spec:
```diff
@@ -105,6 +105,7 @@ BuildRequires: libical-devel
 BuildRequires: ncurses-devel
 BuildRequires: perl
 BuildRequires: postgresql-devel
+BuildRequires: postgresql-contrib
```
Please update pbspro.spec.in as well.

suresh-thelkar replied:
Thanks @mike0042 . I have updated PBSPro.spec.in with the changes that I have done for psbpro.spec in the latest commit.

suresh-thelkar commented:
Codacy has jun run just now. Don't know why it has not run since yesterday. I am looking into these issues and fix them if they are valid.

suresh-thelkar commented:
I have analysed the Codacy issues. All of them are false positive.

subhasisb commented:
looks good Suresh. Just asking, have we remembered to perform valgrind tests on Linux? Also, some tests on Windows would ensure that the changes actually work on windows

subhasisb commented:
What is perplexing me about the codacy warning is why it finds ntohl() okay but not ntohll()

suresh-thelkar commented:
> looks good Suresh. Just asking, have we remembered to perform valgrind tests on Linux? Also, some tests on Windows would ensure that the changes actually work on windows

@subhasisb, Thank you very much for your sign off.  Have ran valgrind on the test suites that are impacted due to these code changes and fixed the issues in the same PR.  Also attached is the log file for Windows.
[WindowsTesting.log](https://github.com/PBSPro/pbspro/files/2648273/WindowsTesting.log)

suresh-thelkar commented:
> What is perplexing me about the codacy warning is why it finds ntohl() okay but not ntohll(). Would adding the required headers for ntohll() per the man page make any difference? I know preprocessor is passing, so the compiler is fine with only netinet.h, but can we try adding the other two:
> #include <sys/types.h>
> #include <inttypes.h>

Thanks for the suggestion. I will add these headers and check if Codacy still throws this warning or not.

suresh-thelkar commented:
@nithinj, I have addressed your comments in the recent commit. Please look into it and provide your sign off or further review comments if any.

@subhasisb, I have made a very few minor changes since your last sign off. Please look into the last commit and let me know if you have any comments. Also I have added the suggested headers to remove the Codacy warning as you suggested.

suresh-thelkar commented:
Squashed and rebased.
